File: flask_backend/routes/employees.py
from flask import Blueprint, request, jsonify, Response
from datetime import datetime
import os

from database.db import get_db, write_db
from utils.crypto import generate_secure_token, build_qr_payload
from utils.qr_generator import generate_qrcode_base64, generate_qrcode_svg

employees_bp = Blueprint('employees', __name__)
BASE_URL = os.environ.get('QR_BASE_URL', 'http://localhost:5000')

def sanitize(emp):
    if not emp: return None
    return {k: v for k, v in emp.items() if k != 'secure_token'}

def next_id(arr):
    return max([e.get('id', 0) for e in arr], default=0) + 1

@employees_bp.route('/', methods=['GET'])
def get_employees():
    try:
        db = get_db()
        search = request.args.get('search', '').lower()
        department = request.args.get('department')
        
        emp_list = list(db['employees'])
        
        if search:
            emp_list = [e for e in emp_list if 
                        search in e.get('name', '').lower() or
                        search in e.get('employee_id', '').lower() or
                        search in e.get('designation', '').lower() or
                        search in e.get('department', '').lower()]
                        
        if department:
            emp_list = [e for e in emp_list if e.get('department') == department]
            
        today = datetime.now().strftime('%Y-%m-%d')
        attendance_today = [a for a in db['attendance'] if a['date'] == today]
        
        result = []
        for e in emp_list:
            sanitized = sanitize(e)
            record = next((a for a in attendance_today if a['employee_id'] == e['employee_id']), None)
            sanitized['present_today'] = bool(record)
            if record:
                is_check_in = (len(record.get('scans', [])) % 2 != 0) if 'scans' in record else not record.get('time_out')
                sanitized['attendance_status'] = 'Checked In' if is_check_in else 'Checked Out'
            else:
                sanitized['attendance_status'] = 'Pending'
            result.append(sanitized)
            
        return jsonify({"success": True, "data": result, "count": len(result)})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to fetch employees"}), 500

@employees_bp.route('/departments', methods=['GET'])
def get_departments():
    db = get_db()
    depts = sorted(list(set(e.get('department') for e in db['employees'] if e.get('department'))))
    return jsonify({"success": True, "data": depts})

@employees_bp.route('/<string:emp_id>', methods=['GET'])
def get_employee(emp_id):
    db = get_db()
    emp = next((e for e in db['employees'] if e.get('employee_id') == emp_id), None)
    if not emp:
        return jsonify({"success": False, "message": "Employee not found"}), 404
    return jsonify({"success": True, "data": sanitize(emp)})

@employees_bp.route('/', methods=['POST'])
def create_employee():
    try:
        db = get_db()
        data = request.json
        required = ['employee_id', 'name', 'designation', 'department', 'dob', 'age', 'gender', 'blood_group', 'address', 'phone', 'emergency_contact']
        
        missing = [k for k in required if data.get(k) in (None, '')]
        if missing:
            return jsonify({"success": False, "message": f"Missing: {', '.join(missing)}"}), 400
            
        if any(e.get('employee_id') == data['employee_id'] for e in db['employees']):
            return jsonify({"success": False, "message": "Employee ID already exists"}), 409
            
        secure_token = generate_secure_token()
        
        github_url = os.environ.get('GITHUB_PAGES_URL', 'https://weguideai.github.io/employee-profiles')
        profile_url = f"{github_url.rstrip('/')}/{data['name'].strip()}.html"
            
        qr_payload = build_qr_payload(secure_token, profile_url)
        qr_data = generate_qrcode_base64(qr_payload)
        
        now = datetime.utcnow().isoformat() + 'Z'
        
        new_emp = {
            "id": next_id(db['employees']),
            "employee_id": data['employee_id'],
            "name": data['name'],
            "designation": data['designation'],
            "department": data['department'],
            "dob": data['dob'],
            "age": int(data['age']),
            "gender": data['gender'],
            "blood_group": data['blood_group'],
            "address": data['address'],
            "phone": data['phone'],
            "emergency_contact": data['emergency_contact'],
            "secure_token": secure_token,
            "profile_url": profile_url,
            "qr_data": qr_data,
            "created_at": now,
            "updated_at": now,
        }
        
        db['employees'].append(new_emp)
        write_db()
        
        return jsonify({"success": True, "data": sanitize(new_emp), "message": "Employee created successfully"}), 201
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to create employee"}), 500

@employees_bp.route('/<string:emp_id>', methods=['PUT'])
def update_employee(emp_id):
    try:
        db = get_db()
        idx = next((i for i, e in enumerate(db['employees']) if e.get('employee_id') == emp_id), -1)
        if idx == -1:
            return jsonify({"success": False, "message": "Employee not found"}), 404
            
        emp = db['employees'][idx]
        data = request.json
        fields = ['name','designation','department','dob','age','gender','blood_group','address','phone','emergency_contact']
        for f in fields:
            if f in data:
                emp[f] = int(data[f]) if f == 'age' else data[f]
                
        if 'profile_url' in data and data['profile_url'] != emp.get('profile_url'):
            emp['profile_url'] = data['profile_url']
            qr_payload = build_qr_payload(emp['secure_token'], emp['profile_url'])
            emp['qr_data'] = generate_qrcode_base64(qr_payload)
                
        emp['updated_at'] = datetime.utcnow().isoformat() + 'Z'
        write_db()
        return jsonify({"success": True, "data": sanitize(emp), "message": "Employee updated successfully"})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to update employee"}), 500

@employees_bp.route('/<string:emp_id>', methods=['DELETE'])
def delete_employee(emp_id):
    try:
        db = get_db()
        before = len(db['employees'])
        db['employees'] = [e for e in db['employees'] if e.get('employee_id') != emp_id]
        db['attendance'] = [a for a in db['attendance'] if a.get('employee_id') != emp_id]
        
        if len(db['employees']) == before:
            return jsonify({"success": False, "message": "Employee not found"}), 404
            
        write_db()
        return jsonify({"success": True, "message": "Employee deleted successfully"})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to delete employee"}), 500

@employees_bp.route('/<string:emp_id>/qr', methods=['GET'])
def get_qr(emp_id):
    try:
        db = get_db()
        emp = next((e for e in db['employees'] if e.get('employee_id') == emp_id), None)
        if not emp:
            return jsonify({"success": False, "message": "Employee not found"}), 404
            
        if request.args.get('format') == 'svg':
            svg = generate_qrcode_svg(build_qr_payload(emp['secure_token'], emp['profile_url']))
            return Response(svg, mimetype='image/svg+xml', headers={'Content-Disposition': f'attachment; filename="qr_{emp_id}.svg"'})
            
        return jsonify({"success": True, "data": {"qr_base64": emp['qr_data'], "employee_id": emp_id, "name": emp['name']}})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to get QR"}), 500

@employees_bp.route('/<string:emp_id>/attendance', methods=['GET'])
def get_employee_attendance(emp_id):
    try:
        db = get_db()
        month = request.args.get('month')
        year = request.args.get('year')
        logs = [a for a in db['attendance'] if a.get('employee_id') == emp_id]
        
        if month and year:
            logs = [a for a in logs if a['date'].startswith(f"{year}-{int(month):02d}")]
            
        logs.sort(key=lambda x: x['date'], reverse=True)
        return jsonify({"success": True, "data": logs, "count": len(logs)})
    except Exception as e:
        print(e)
        return jsonify({"success": False, "message": "Failed to fetch attendance"}), 500

File: backend/routes/employees.js
const express = require('express');
const router = express.Router();
const { getDB } = require('../database/db');
const { generateSecureToken, buildQrPayload } = require('../utils/crypto');
const { generateQRCodeBase64, generateQRCodeSVG } = require('../utils/qrGenerator');

const BASE_URL = process.env.QR_BASE_URL || 'http://localhost:5000';

// ─── Helper: strip secure_token from output ───────────────────────────────────
function sanitize(emp) {
  if (!emp) return null;
  const { secure_token, ...rest } = emp;
  return rest;
}

// ─── GET /api/employees ────────────────────────────────────────────────────────
router.get('/', (req, res) => {
  try {
    const db = getDB();
    const { search, department } = req.query;

    let sql = 'SELECT * FROM employees WHERE 1=1';
    const params = [];

    if (search) {
      sql += ` AND (name LIKE ? OR employee_id LIKE ? OR designation LIKE ? OR department LIKE ?)`;
      const s = `%${search}%`;
      params.push(s, s, s, s);
    }
    if (department) {
      sql += ' AND department = ?';
      params.push(department);
    }
    sql += ' ORDER BY created_at DESC';

    let list = db.prepare(sql).all(...params);

    const today = new Date().toISOString().split('T')[0];
    const todayAttendance = db.prepare('SELECT * FROM attendance WHERE date = ?').all(today);

    const result = list.map(e => {
      const s = sanitize(e);
      const record = todayAttendance.find(a => a.employee_id === e.employee_id);
      s.present_today = !!record;
      if (record) {
        const scans = JSON.parse(record.scans || '[]');
        const isCheckIn = scans.length % 2 !== 0;
        s.attendance_status = isCheckIn ? 'Checked In' : 'Checked Out';
      } else {
        s.attendance_status = 'Pending';
      }
      return s;
    });

    res.json({ success: true, data: result, count: result.length });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: 'Failed to fetch employees' });
  }
});

// ─── GET /api/employees/departments ───────────────────────────────────────────
router.get('/departments', (req, res) => {
  const db = getDB();
  const rows = db.prepare('SELECT DISTINCT department FROM employees ORDER BY department').all();
  res.json({ success: true, data: rows.map(r => r.department) });
});

// ─── GET /api/employees/:id ────────────────────────────────────────────────────
router.get('/:id', (req, res) => {
  const db = getDB();
  const emp = db.prepare('SELECT * FROM employees WHERE employee_id = ?').get(req.params.id);
  if (!emp) return res.status(404).json({ success: false, message: 'Employee not found' });
  res.json({ success: true, data: sanitize(emp) });
});

// ─── POST /api/employees ───────────────────────────────────────────────────────
router.post('/', async (req, res) => {
  try {
    const db = getDB();
    const {
      employee_id, name, designation, department,
      dob, age, gender, blood_group,
      address, phone, emergency_contact
    } = req.body;

    const required = { employee_id, name, designation, department, dob, age, gender, blood_group, address, phone, emergency_contact };
    const missing = Object.entries(required).filter(([, v]) => v === undefined || v === '');
    if (missing.length) {
      return res.status(400).json({ success: false, message: `Missing: ${missing.map(([k]) => k).join(', ')}` });
    }

    const existing = db.prepare('SELECT id FROM employees WHERE employee_id = ?').get(employee_id);
    if (existing) {
      return res.status(409).json({ success: false, message: 'Employee ID already exists' });
    }

    const secure_token = generateSecureToken();
    const github_url   = process.env.GITHUB_PAGES_URL || 'https://weguideai.github.io/employee-profiles';
    const profile_url  = `${github_url.replace(/\/$/, '')}/${name.trim()}.html`;
    const qrPayload    = buildQrPayload(secure_token, profile_url);
    const qr_data      = await generateQRCodeBase64(qrPayload);
    const now          = new Date().toISOString();

    db.prepare(`
      INSERT INTO employees
        (employee_id, name, designation, department, dob, age, gender, blood_group,
         address, phone, emergency_contact, secure_token, profile_url, qr_data, created_at, updated_at)
      VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
    `).run(employee_id, name, designation, department, dob, parseInt(age), gender, blood_group,
           address, phone, emergency_contact, secure_token, profile_url, qr_data, now, now);

    const newEmp = db.prepare('SELECT * FROM employees WHERE employee_id = ?').get(employee_id);
    res.status(201).json({ success: true, data: sanitize(newEmp), message: 'Employee created successfully' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: 'Failed to create employee' });
  }
});

// ─── PUT /api/employees/:id ────────────────────────────────────────────────────
router.put('/:id', async (req, res) => {
  try {
    const db = getDB();
    const emp = db.prepare('SELECT * FROM employees WHERE employee_id = ?').get(req.params.id);
    if (!emp) return res.status(404).json({ success: false, message: 'Employee not found' });

    const fields = ['name','designation','department','dob','age','gender','blood_group','address','phone','emergency_contact'];
    const updates = {};
    fields.forEach(f => {
      if (req.body[f] !== undefined) {
        updates[f] = f === 'age' ? parseInt(req.body[f]) : req.body[f];
      }
    });

    if (req.body.profile_url !== undefined && req.body.profile_url !== emp.profile_url) {
      updates.profile_url = req.body.profile_url;
      const qrPayload = buildQrPayload(emp.secure_token, req.body.profile_url);
      updates.qr_data = await generateQRCodeBase64(qrPayload);
    }

    updates.updated_at = new Date().toISOString();

    const setClauses = Object.keys(updates).map(k => `${k} = ?`).join(', ');
    const values = Object.values(updates);
    db.prepare(`UPDATE employees SET ${setClauses} WHERE employee_id = ?`).run(...values, req.params.id);

    const updated = db.prepare('SELECT * FROM employees WHERE employee_id = ?').get(req.params.id);
    res.json({ success: true, data: sanitize(updated), message: 'Employee updated successfully' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: 'Failed to update employee' });
  }
});

// ─── DELETE /api/employees/:id ─────────────────────────────────────────────────
router.delete('/:id', (req, res) => {
  try {
    const db = getDB();
    const info = db.prepare('DELETE FROM employees WHERE employee_id = ?').run(req.params.id);
    if (info.changes === 0) {
      return res.status(404).json({ success: false, message: 'Employee not found' });
    }
    db.prepare('DELETE FROM attendance WHERE employee_id = ?').run(req.params.id);
    res.json({ success: true, message: 'Employee deleted successfully' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: 'Failed to delete employee' });
  }
});

// ─── GET /api/employees/:id/qr ─────────────────────────────────────────────────
router.get('/:id/qr', async (req, res) => {
  try {
    const db = getDB();
    const emp = db.prepare('SELECT * FROM employees WHERE employee_id = ?').get(req.params.id);
    if (!emp) return res.status(404).json({ success: false, message: 'Employee not found' });

    if (req.query.format === 'svg') {
      const svg = await generateQRCodeSVG(buildQrPayload(emp.secure_token, emp.profile_url));
      res.setHeader('Content-Type', 'image/svg+xml');
      res.setHeader('Content-Disposition', `attachment; filename="qr_${emp.employee_id}.svg"`);
      return res.send(svg);
    }

    res.json({ success: true, data: { qr_base64: emp.qr_data, employee_id: emp.employee_id, name: emp.name } });
  } catch (err) {
    res.status(500).json({ success: false, message: 'Failed to get QR' });
  }
});

// ─── GET /api/employees/:id/attendance ────────────────────────────────────────
router.get('/:id/attendance', (req, res) => {
  try {
    const db = getDB();
    const { month, year } = req.query;
    let sql = 'SELECT * FROM attendance WHERE employee_id = ?';
    const params = [req.params.id];

    if (month && year) {
      sql += ` AND strftime('%m', date) = ? AND strftime('%Y', date) = ?`;
      params.push(String(month).padStart(2, '0'), String(year));
    }
    sql += ' ORDER BY date DESC';

    const logs = db.prepare(sql).all(...params).map(r => ({
      ...r, scans: JSON.parse(r.scans || '[]')
    }));
    res.json({ success: true, data: logs, count: logs.length });
  } catch (err) {
    res.status(500).json({ success: false, message: 'Failed to fetch attendance' });
  }
});

module.exports = router;
